### ISSUE TITLE
Rule proposal: Disallow calling add/removeEventListener with the result of a bound function call

### ISSUE THREAD

**Issue by bschlenk** [sindresorhus/eslint-plugin-unicorn#682]
Calling `this.addEventListener('event', myFunc.bind(this))` and `this.removeEventListener('event', myFunc.bind(this))` has always been an indication of an error to me. The corresponding `removeEventListener` isn't actually doing anything because the `.bind` call creates a new function.

Not sure what I'd call this rule. `disallow-anonymous-event-handler` ?

This was [requested in eslint](https://github.com/eslint/eslint/issues/6151) but for some reason got rejcted.

## Fail

```js
class MyClass extends HTMLElement {
  connectedCallback() {
    this.addEventListener('click', this.handleClick.bind(this));
  }

  disconnectedCallback() {
    this.removeEventListener('click', this.handleClick.bind(this));
  }

  handleClick() { /* ... */ }
}
```

## Pass

```js
class MyClass extends HTMLElement {
  connectedCallback() {
    this.addEventListener('click', this.handleClick);
  }

  disconnectedCallback() {
    this.removeEventListener('click', this.handleClick);
  }

  handleClick = () => { /* ... */ }
}
```

**Comment by sindresorhus** [comment]
I don't really see why we should disallow `addEventListener`. Disallowing using `.bind` in `removeEventListener` totally makes sense though. The rule name should be focused on that.

**Comment by fisker** [comment]
Is this even valid code?

```js
this.removeEventListener('click', this.handleClick.bind(this));
```

How to remove listener with a total new function

**Comment by bschlenk** [comment]
It’s valid code in that it doesn’t throw any errors, but it doesn’t remove the event listener. That would be the reason to have this rule. 

What about this case:

```
this.addEventListener(this.handleClick.bind(this);
this.removeEventListener(this.handleClick)
```

I’ve seen someone write this code too. The `removeEventListener` call isn’t using bind, but it is still broken.

Maybe it can be configurable which methods it disallows? I’d set it up to fail on both, since there are plenty of alternatives to `bind` that work. But someone else can set it up only to fail for `removeEventListener`. The point is to prevent code that can easily cause errors.

**Comment by sindresorhus** [comment]
> Maybe it can be configurable which methods it disallows? I’d set it up to fail on both, since there are plenty of alternatives to bind that work. But someone else can set it up only to fail for removeEventListener. The point is to prevent code that can easily cause errors.

I prefer to keep it opinionated unless there are good arguments for adding options.

**Comment by sindresorhus** [comment]
This rule is accepted.

**Comment by fisker** [comment]
I don't really see the point forbid `.addEventListener('click', foo.bind(this))`, if your `.removeEventListener` don't work as expect, you should check your logic, not relay on the static check.

I don't think we can track the same object called `removeEventListener`.

I'm okay with `.removeEventListener('click', foo.bind(this))`, because it's obviously useless code.

How about forbid `FunctionExpression`, `ArrowFunctionExpression`, and `function#bind` in `.removeEventListener` only.

If you still want forbid `.addEventListener`, I suggest you can use [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax), like

```js
"no-restricted-syntax": [
	"error", 
	'CallExpression[callee.property.name="addEventListener"][arguments.1.callee.property.name="bind"]'
]
```
(Not strict, but should enough)

**Comment by bschlenk** [comment]
I can get behind that, I'll focus on only checking the `.removeEventListener` calls. The `no-restricted-syntax` setup for `.addEventListener` seems kinda hacky but it would work.

My main argument for optionally allowing the rule to check both is that you might need to add the `.removeEventListener` later on, and not realize that `.addEventListener` was called with a bound function. Disallowing it from the start prevents this kind of issue. Testing that `.removeEventListener` worked is not always easy, since a lot of times it is there just to prevent memory leaks.

**Comment by fisker** [comment]
How about `no-invalid-event-listener-removal`?

**Comment by fisker** [comment]
We'll forbid any listener that it's not a reference, including

```js
foo.removeEventListener('click', function () {})
```

```js
foo.removeEventListener('click', () => {})
```

```js
foo.removeEventListener('click', bar.bind(null))
```

**Comment by bschlenk** [comment]
I think I'd go with `no-invalid-remove-event-listener` so that it's close to the actual method name. I'm going to take a stab at implementing this. Thanks for the discussion!

**Comment by fisker** [comment]
> I think I'd go with no-invalid-remove-event-listener 

Sounds good.
